src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,889 @@
+
+  bool successful() const { return _error == NO_ERROR; }
+
+  JfrSampleType type() const { return _type; }
+
+  JfrTicks start_time() const { return _start_time; }
+  void set_end_time(JfrTicks end_time) { _end_time = end_time; }
+  JfrTicks end_time() const { return _end_time; }
+  void set_sampled_thread(JavaThread* thread) { _sampled_thread = thread; }
+  JavaThread* sampled_thread() const { return _sampled_thread; }
+
+  JfrAsyncStackTrace& stacktrace() { return _stacktrace; }
+
+  int64_t sampling_period() const { return _sampling_period; }
+
+  enum SampleError {
+    NO_ERROR = 0,
+    ERROR_NO_TRACE = 1,
+    ERROR_NO_TOPFRAME = 2,
+    ERROR_JAVA_WALK_FAILED = 3,
+    ERROR_NATIVE_WALK_FAILED = 4,
+    ERROR_NO_TOP_METHOD = 5,
+    ERROR_NO_LAST_JAVA_FRAME = 6
+  };
+
+  // Record a trace of the current thread
+  void record_trace(JavaThread* jt, void* ucontext, int64_t sampling_period) {
+    _stacktrace = JfrAsyncStackTrace(_frames, _max_frames);
+    set_sampled_thread(jt);
+    _type = NO_SAMPLE;
+    _error = ERROR_NO_TRACE;
+    _start_time = _end_time = JfrTicks::now();
+    _sampling_period = sampling_period;
+    if (!jt->in_deopt_handler() && !Universe::heap()->is_stw_gc_active())  {
+      ThreadInAsgct tia(jt);
+      if (thread_state_in_java(jt)) {
+        record_java_trace(jt, ucontext);
+      } else if (thread_state_in_native(jt)) {
+        record_native_trace(jt, ucontext);
+      }
+    }
+    _end_time = JfrTicks::now();
+  }
+
+private:
+
+  void record_java_trace(JavaThread* jt, void* ucontext) {
+    _type = JAVA_SAMPLE;
+    JfrGetCallTrace trace(true, jt);
+    frame topframe;
+    if (trace.get_topframe(ucontext, topframe)) {
+      _error = _stacktrace.record_async(jt, topframe) ? NO_ERROR : ERROR_JAVA_WALK_FAILED;
+    }
+  }
+
+  void record_native_trace(JavaThread* jt, void* ucontext) {
+    // When a thread is only attach it will be native without a last java frame
+   _type = NATIVE_SAMPLE;
+    _error = ERROR_NO_TRACE;
+    if (!jt->has_last_Java_frame()) {
+      _error = ERROR_NO_LAST_JAVA_FRAME;
+      return;
+    }
+    frame topframe;
+    if (!jt->pd_get_top_frame_for_signal_handler(&topframe, ucontext, false)) {
+      _error = ERROR_NO_TOPFRAME;
+      return;
+    }
+    frame first_java_frame;
+    Method* method = nullptr;
+    JfrGetCallTrace gct(false, jt);
+    if (!gct.find_top_frame(topframe, &method, first_java_frame)) {
+      _error = ERROR_NO_TOPFRAME;
+      return;
+    }
+    if (method == nullptr) {
+      _error = ERROR_NO_TOP_METHOD;
+      return;
+    }
+    topframe = first_java_frame;
+    _error = _stacktrace.record_async(jt, topframe) ? NO_ERROR: ERROR_NATIVE_WALK_FAILED;
+  }
+};
+
+// Fixed size array-based MPMC queue for storing traces.
+// Does not own any frames
+class JfrTraceQueue {
+
+  struct Element {
+    // Encodes generation of the element (to solve ABA problem)
+    // along with full/empty flag in the highest bit
+    u4 _state;
```

Comment:
Can you add more commentary explaining the operation of the queue please and how this "state" interacts with enqueue/dequeue.

Reply:
@jbachorik and @apangin: can you take this?